auth/src/main/java/com/firebase/ui/auth/viewmodel/phone/PhoneProviderResponseHandler.java

```diff
@@ -31,14 +35,23 @@ public void startSignIn(@NonNull PhoneAuthCredential credential,
         }
         setResult(Resource.<IdpResponse>forLoading());
 
-        getAuth().signInWithCredential(credential)
-                .addOnCompleteListener(new OnCompleteListener<AuthResult>() {
+        AnonymousUpgradeUtils.signInAndLinkWithPhoneNumber(getAuth(), getArguments(), credential)
+                .addOnSuccessListener(new OnSuccessListener<AuthResult>() {
                     @Override
-                    public void onComplete(@NonNull Task<AuthResult> task) {
-                        if (task.isSuccessful()) {
-                            handleSuccess(response, task.getResult());
+                    public void onSuccess(AuthResult result) {
+                        handleSuccess(response, result);
+                    }
+                })
+                .addOnFailureListener(new OnFailureListener() {
```

Comment:
optional: if you look at my PR you'll notice I made a type of `Task` that uses the type system to force you to handle merge exceptions at all call sites.  Since you are pretty constantly doing `instanceof` checks on the exceptions for these operations you may want to consider something like that.

Reply:
I'll come back to this.